test/e2e_node/e2e_node_suite_test.go

```diff
@@ -100,6 +103,24 @@ func registerNodeFlags(flags *flag.FlagSet) {
 	flag.Var(cliflag.NewMapStringString(&framework.TestContext.RuntimeConfig), "runtime-config", "The runtime configuration used on node e2e tests.")
 	flags.BoolVar(&framework.TestContext.RequireDevices, "require-devices", false, "If true, require device plugins to be installed in the running environment.")
 	flags.Var(cliflag.NewMapStringBool(&featureGates), "feature-gates", "A set of key=value pairs that describe feature gates for alpha/experimental features.")
+
+	if framework.TestContext.Provider == "" && gcemetadata.OnGCE() {
+		gceZone, err := gcemetadata.Zone()
+		if err != nil {
+			klog.Exitf("Failed to get GCE zone: %v", err)
+		}
+
+		gceProject, err := gcemetadata.ProjectID()
+		if err != nil {
+			klog.Exitf("Failed to get GCE projectID: %v", err)
+		}
+
+		flag.StringVar(&framework.TestContext.CloudConfig.Zone, "gce-zone", gceZone, "GCE zone being used, if applicable")
+		flag.StringVar(&framework.TestContext.CloudConfig.Zone, "gce-project", gceProject, "GCE project being used, if applicable")
+
+		flag.StringVar(&framework.TestContext.Provider, "provider", "gce", "The name of the Kubernetes provider (gce, none, etc). Defaults to GCE if unset and running on GCE VM.")
```

Comment:
how are these three fields used in e2e_node tests?

is there a reason not to register them unconditionally with non-gce-specific names and empty default values, and expect the invocation in a cloud provider environment to do the work to figure out the zone/project?

Reply:
This still WIP, the overall issue I'm trying to address here is there are some tests which are run as both cluster and node e2e tests, and they are only supported in GCE.

The node e2e runner currently is not aware it is running in GCE, so a call like https://github.com/kubernetes/kubernetes/blob/6340b19ddac0fb64e13cc54987d001294ef26ab3/test/e2e/common/node/runtimeclass.go#L86

results in the test being skipped in node e2e (even if it's running in GCE). Setting these flags overrides the test provider when it running in GCE. The zone and project are actually unused, but it seems like it's needed for the gce provider to initialize

https://github.com/kubernetes/kubernetes/blob/6340b19ddac0fb64e13cc54987d001294ef26ab3/test/e2e/framework/providers/gce/gce.go#L67-L71

Another solution here to avoid adding GCE dependency is to change the test logic to avoid skipping on non GCE and instead look at per environment flag that could be in the test.